### PR TITLE
refactor(gossipsub): remove `derive-builder` dev-dependency

### LINKED ISSUE #3228: gossipsub/tests: Remove `derive-builder` dependency
Let's try to remove the `derive-builder` dependency in `libp2p-gossipsub`. It is only used in tests and thus likely doesn't warrant a dedicated dependency.

https://github.com/libp2p/rust-libp2p/issues/3177

### PR DESCRIPTION
**Pull request by StemCll**
## Description

Remove the `derive_builder` dev-dependency in gossipsub. We can manually implement the builder functionality on top of the `Default` instance of `InjectNodes`.

Resolved #3228.

## Links to any relevant issues

[Related issue](https://github.com/libp2p/rust-libp2p/issues/3228)

## Change checklist

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: protocols/gossipsub/src/behaviour/tests.rs [hunk 1]
```diff
@@ -38,13 +38,12 @@ use std::hash::{Hash, Hasher};
...
 
-#[derive(Default, Builder, Debug)]
-#[builder(default)]
+#[derive(Default, Debug)]
 struct InjectNodes<D, F>
 // TODO: remove trait bound Default when this issue is fixed:
 //  https://github.com/colin-kiegel/rust-derive-builder/issues/93
 where
-    D: DataTransform + Default + Clone + Send + 'static,
+    D: DataTransform + Clone + Default + Send + 'static,
```
Comment (thomaseizinger) [inline protocols/gossipsub/src/behaviour/tests.rs]: Please minimize the diff as much as possible :)

Swapping these seems unnecessary.

File: protocols/gossipsub/src/behaviour/tests.rs [hunk 2]
```diff
@@ -60,30 +59,30 @@ where
 
 impl<D, F> InjectNodes<D, F>
 where
-    D: DataTransform + Default + Clone + Send + 'static,
+    D: DataTransform + Clone + Default + Send + 'static,
     F: TopicSubscriptionFilter + Clone + Default + Send + 'static,
 {
-    pub fn create_network(self) -> (Gossipsub<D, F>, Vec<PeerId>, Vec<TopicHash>) {
+    pub fn create_network(&self) -> (Gossipsub<D, F>, Vec<PeerId>, Vec<TopicHash>) {
```
Comment (thomaseizinger) [inline protocols/gossipsub/src/behaviour/tests.rs]: Is this signature change necessary? What is the point of keeping the `InjectNodes` builder around? Unless I am missing something, consuming a builder is idiomatic.

**Review by thomaseizinger — CHANGES_REQUESTED** [review]
Direction looks good to me, thank you!

Two requests.

File: protocols/gossipsub/src/behaviour/tests.rs (hunk 1 above)
Comment (StemCll) [reply protocols/gossipsub/src/behaviour/tests.rs]: Done :heavy_check_mark:

File: protocols/gossipsub/src/behaviour/tests.rs (hunk 2 above)
Comment (StemCll) [reply protocols/gossipsub/src/behaviour/tests.rs]: Done :heavy_check_mark:

File: protocols/gossipsub/src/behaviour/tests.rs [hunk 3]
```diff
@@ -108,32 +107,62 @@ where
...
+    InjectNodes::default()
 }
 
-fn inject_nodes1() -> InjectNodesBuilder<IdentityTransform, AllowAllSubscriptionFilter> {
-    inject_nodes()
+fn inject_nodes1() -> InjectNodes<IdentityTransform, AllowAllSubscriptionFilter> {
+    InjectNodes::<IdentityTransform, AllowAllSubscriptionFilter>::default()
 }
 
-// helper functions for testing
```
Comment (thomaseizinger) [inline protocols/gossipsub/src/behaviour/tests.rs:135]: Nit: Can we retain the position of this comment?

**Review by thomaseizinger — CHANGES_REQUESTED** [review]
One more request and please address clippy failures: "clippy (stable)" job only! :)

File: protocols/gossipsub/src/behaviour/tests.rs (hunk 3 above)
Comment (StemCll) [reply protocols/gossipsub/src/behaviour/tests.rs:135]: :heavy_check_mark:

File: core/src/transport/memory.rs [hunk 4]
```diff
@@ -227,7 +227,8 @@ impl Transport for MemoryTransport {
             return Err(TransportError::MultiaddrNotSupported(addr));
         };
 
-        DialFuture::new(port).ok_or(TransportError::Other(MemoryTransportError::Unreachable))
+        DialFuture::new(port)
+            .ok_or_else(|| TransportError::Other(MemoryTransportError::Unreachable))
```
Comment (thomaseizinger) [inline core/src/transport/memory.rs]: Did clippy actually suggest these? That is new to me. Normally it recommends the opposite, i.e. if there is no allocation, use `ok_or`.

Comment (thomaseizinger) [reply core/src/transport/memory.rs]: Also, this isn't flagged in other PRs. Can you please revert this (and others)?

File: protocols/gossipsub/src/behaviour/tests.rs [hunk 5]
```diff
@@ -108,28 +107,58 @@ where
...
+        self.peer_no = peer_no;
+        self
+    }
+
+    fn topics(mut self, topics: Vec<String>) -> Self {
+        self.topics = topics;
+        self
+    }
+
+    fn set_to_subscribe(mut self, to_subscribe: bool) -> Self {
```
Comment (thomaseizinger) [inline protocols/gossipsub/src/behaviour/tests.rs]: I am guessing that clippy made you rename this because it doesn't follow some convention?

Sorry to go back on this but I'd prefer all builder functions to be consistent and just slap a `#[allow(clippy::lint_name)]` on here. Otherwise we have this unnecessary diff when all we wanted was to replace the macro :)

**Review by thomaseizinger — CHANGES_REQUESTED** [review]
Thanks for pushing this forward, some questions.

Also, please don't force-push. We squash merge in this repository :)

File: core/src/transport/memory.rs (hunk 4 above)
Comment (StemCll) [reply core/src/transport/memory.rs]: Yes, I've had wrong toolchain when running clippy. Reverting these changes.

**Review by thomaseizinger — APPROVED** [review]
Nice, thank you for working with me on this! Appreciate that we could minimize the diff :)

**Comment by StemCll** [comment]
Sure! Moving to the next issue :)
